README.md
# Latex to Markdown Compiler

<img src="https://raw.githubusercontent.com/Blyxyas/Latex-to-Markdown-Compiler/master/images/logo.png" />
This is a latex to markdown compiler.

## Usage

Open the target file, open the command palette (With *<kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>P</kbd>* In Windows & Linux, or *<kbd>⌘</kbd> + <kbd>Shift</kbd> + <kbd>P</kbd>* for MacOS), and select Search for **`Compile Latex`**. That's it! It will replace all you latex formulas with the HTML tags that will render them.

## Features

<small>(All this formulas were compiled using the compiler)</small>
Use `$` to open an inline math formula. Like this!

(View the source for this formulas in the [Latex file](https://github.com/Blyxyas/Latex-to-Markdown-Compiler/blob/master/formulas.tex))

This is an inline formula <img src="https://render.githubusercontent.com/render/math?math=\bbox[%230d1117]{\color{%23fff}{E%3Dmc%5E2}}" /> which means that I can keep writing in this line.

You can also use `$$` to open a block math formula.
<h3 align="center"><img src="https://render.githubusercontent.com/render/math?math=\bbox[%230d1117]{\color{%23fff}{F_g%3D%5Cfrac%7BGm_1m_2%7D%7Bd%5E2%7D}}" /></h3>



The blocks are centered and are rendered with a `h3` tag.

## Installation

You can install it from the Command palette using <kbd>Command</kbd> + <kbd>P</kbd> and writing `ext install Blyxyas.latexcompilar`, You can also search for `Blyxyas.latexcompiler` in the integrated Marketplace in Visual Studio Code. **(You can also install it in Visual Studio)**.

### Building it yourself

**You can get the latest version installing it with a VSIX file.**
This is not recommended, see [Installation](#Installation) to install it the recommended way.

1. Clone the repo


```bash
git clone https://github.com/Blyxyas/Latex-to-Markdown-Compiler.git
```

2. Install the `vsce` CLI tool ([Node.js](https://nodejs.org/en/) required)

```bash
npm install -g vsce
```

3. Build the package


```bash
vsce package
```

4. Install it from VSIX

-  1. Open Command Palette (<kbd>Command</kbd> + <kbd>Shift</kbd> + <kbd>P</kbd>)
-  2. Search `Install from VSIX`
-  3. Point to the VSIX file.

Or you can also use the CLI


```bash
code --install-extension PACKAGENAME.vsix
```

\[Replace `PACKAGENAME` to the name of the build you made.]

## Configuration

- `textColor`: Changes the color of the text
- `backgroundColor`: Changes the color of the background

You can use HEX Codes (E.g. `#333`) or a name (e.g. `red`)

## Details

* **You can use `\` to escape the `$` character.**

* The HTML tags for the blocks will be added at the final `$$` of the block. **Everything between the first `$$` and the last `$$` will be removed!**

**Don't do**
```latex
<h3 align="center"><img src="https://render.githubusercontent.com/render/math?math=\bbox[%23101414]{\color{%23fff}{%20BLOCK%20%24%24%60%60%60%2a%2aDo%2a%2a%60%60%60latex}}" /></h3>




BLOCK

```

CHANGELOG.md
# Change Log

All notable changes to the "simplelatex" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/extension.ts
// First time using JS in more than a year and a half, sorry (And first time using TS ever)

"use strict";

import * as vscode from "vscode";
import fs = require("fs");
import { exit } from "process";

export function activate(context: vscode.ExtensionContext) {
	

	function fixedEncodeURIComponent(str: string) {
		return encodeURIComponent(str).replace(/[!'()*]/g, function(c) {
		  return '%' + c.charCodeAt(0).toString(16);
		});
	  }

	//
	// ─── CONFIGURATION ──────────────────────────────────────────────────────────────
	//

// We read the configuration file
	const config = vscode.workspace.getConfiguration("latexcompiler");
	let backgroundColor: string | undefined = config.get("backgroundColor");
	let textColor: string | undefined = config.get("textColor");

	function compileFile(content: string): string {

		let expression: string[] = [];
		let lineArr: string[];
		
		// <======= BLOCKS =======>
		let inBlock: boolean = false;
		let firstBlockIndex: number = 0;
		let lastBlockIndex: number = 0;
		let blockContent: string = "";
		
		// We inspect every line of file
		let lines: string[] = content.split("\n");
		for (let i = 0; i < lines.length; i++) {
			// We check for blocks
			// Delimiters: $ , $$

			// <======== BLOCK EXPRESSIONS =========>	

			if (lines[i].indexOf("$$") !== -1 && lines[i][lines[i].indexOf("$") - 1] !== "\\" && lines[i][lines[i].indexOf("$") - 1] !== "`") {
				inBlock = !inBlock;
				if (inBlock) {
					lines[i] = lines[i].replace("$$", "");
					firstBlockIndex = i;
				}
				else {
					lines[i] = lines[i].replace("$$", "");
					lastBlockIndex = i;

					blockContent = lines.slice(firstBlockIndex, lastBlockIndex + 1).join("");
					lines[firstBlockIndex] = `<h3 align="center"><img src="https://render.githubusercontent.com/render/math?math=${`\\bbox[${backgroundColor?.replace("#", "%23")}]{\\color{${textColor?.replace("#", "%23")}}{${fixedEncodeURIComponent(blockContent)}}}`}" /></h3>`;

					for (let toClear = firstBlockIndex + 1; toClear < lastBlockIndex; toClear++) {
						lines[toClear] = "";
					}
				}
			}

			else {
				if (lines[i].indexOf("$") !== -1 && lines[i].indexOf("$$") === -1 && lines[i][lines[i].indexOf("$") - 1] !== "\\" && lines[i][lines[i].indexOf("$") - 1] !== "`") {
					// We replace the inline expression with the img tag
					lineArr = lines[i].split("$");
					for (let j = 0; j < lineArr.length; j++) {
						if (j % 2 !== 0 && lineArr[j - 1] !== "\\") {
							lineArr[j] = `<img src="https://render.githubusercontent.com/render/math?math=${`\\bbox[${backgroundColor?.replace("#", "%23")}]{\\color{${textColor?.replace("#", "%23")}}{${fixedEncodeURIComponent(lineArr[j])}}}`}" />`;
						};
					};

					// Now we join the array back to a string
					lines[i] = lineArr.join("");
				};
			};

			// We check for inline expressions
		};
		return lines.join("\n");
	};

	// First, we register a command
	let disposable = vscode.commands.registerCommand("latexcompiler.compile", () => {
		// We alert the user
		// Getting the current workspace folder
		if (vscode.workspace.workspaceFolders !== undefined) {
			let wf = vscode.workspace.workspaceFolders[0].uri.path;
			let f = vscode.workspace.workspaceFolders[0].uri.fsPath;
		}

		// Asking the user to specify a path inside the workspace

		if (vscode.window.activeTextEditor !== undefined) {
		vscode.workspace
			.openTextDocument(vscode.window.activeTextEditor.document.uri.fsPath)
			.then((document) => {
				if (document.isDirty) {
					document.save();
				};
				// Now we read the file
				let text = document.getText();
				// Now we write to the file the new content;
				if (vscode.window.activeTextEditor !== undefined) {
					fs.writeFile(vscode.window.activeTextEditor.document.uri.fsPath, compileFile(text), (err) => {
						if (err) {
							let error = vscode.window.createOutputChannel("Latex Compiler Compilation Error");
							error.appendLine("An error ocurred, no more information available.");
							error.show();
							exit();
						}
						else {
							vscode.window.showInformationMessage("File Compiled!");
						}
					});
				}
			});
		}
	});
}

export function deactivate() {}


out/extension.js
// First time using JS in more than a year and a half, sorry (And first time using TS ever)
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deactivate = exports.activate = void 0;
const vscode = require("vscode");
const fs = require("fs");
const process_1 = require("process");
function activate(context) {
    function fixedEncodeURIComponent(str) {
        return encodeURIComponent(str).replace(/[!'()*]/g, function (c) {
            return '%' + c.charCodeAt(0).toString(16);
        });
    }
    //
    // ─── CONFIGURATION ──────────────────────────────────────────────────────────────
    //
    // We read the configuration file
    const config = vscode.workspace.getConfiguration("latexcompiler");
    let backgroundColor = config.get("backgroundColor");
    let textColor = config.get("textColor");
    function compileFile(content) {
        let expression = [];
        let lineArr;
        // <======= BLOCKS =======>
        let inBlock = false;
        let firstBlockIndex = 0;
        let lastBlockIndex = 0;
        let blockContent = "";
        // We inspect every line of file
        let lines = content.split("\n");
        for (let i = 0; i < lines.length; i++) {
            // We check for blocks
            // Delimiters: $ , $$
            // <======== BLOCK EXPRESSIONS =========>	
            if (lines[i].indexOf("$$") !== -1 && lines[i][lines[i].indexOf("$") - 1] !== "\\" && lines[i][lines[i].indexOf("$") - 1] !== "`") {
                inBlock = !inBlock;
                if (inBlock) {
                    lines[i] = lines[i].replace("$$", "");
                    firstBlockIndex = i;
                }
                else {
                    lines[i] = lines[i].replace("$$", "");
                    lastBlockIndex = i;
                    blockContent = lines.slice(firstBlockIndex, lastBlockIndex + 1).join("");
                    lines[firstBlockIndex] = `<h3 align="center"><img src="https://render.githubusercontent.com/render/math?math=${`\\bbox[${backgroundColor?.replace("#", "%23")}]{\\color{${textColor?.replace("#", "%23")}}{${fixedEncodeURIComponent(blockContent)}}}`}" /></h3>`;
                    for (let toClear = firstBlockIndex + 1; toClear < lastBlockIndex; toClear++) {
                        lines[toClear] = "";
                    }
                }
            }
            else {
                if (lines[i].indexOf("$") !== -1 && lines[i].indexOf("$$") === -1 && lines[i][lines[i].indexOf("$") - 1] !== "\\" && lines[i][lines[i].indexOf("$") - 1] !== "`") {
                    // We replace the inline expression with the img tag
                    lineArr = lines[i].split("$");
                    for (let j = 0; j < lineArr.length; j++) {
                        if (j % 2 !== 0 && lineArr[j - 1] !== "\\") {
                            lineArr[j] = `<img src="https://render.githubusercontent.com/render/math?math=${`\\bbox[${backgroundColor?.replace("#", "%23")}]{\\color{${textColor?.replace("#", "%23")}}{${fixedEncodeURIComponent(lineArr[j])}}}`}" />`;
                        }
                        ;
                    }
                    ;
                    // Now we join the array back to a string
                    lines[i] = lineArr.join("");
                }
                ;
            }
            ;
            // We check for inline expressions
        }
        ;
        return lines.join("\n");
    }
    ;
    // First, we register a command
    let disposable = vscode.commands.registerCommand("latexcompiler.compile", () => {
        // We alert the user
        // Getting the current workspace folder
        if (vscode.workspace.workspaceFolders !== undefined) {
            let wf = vscode.workspace.workspaceFolders[0].uri.path;
            let f = vscode.workspace.workspaceFolders[0].uri.fsPath;
        }
        // Asking the user to specify a path inside the workspace
        if (vscode.window.activeTextEditor !== undefined) {
            vscode.workspace
                .openTextDocument(vscode.window.activeTextEditor.document.uri.fsPath)
                .then((document) => {
                if (document.isDirty) {
                    document.save();
                }
                ;
                // Now we read the file
                let text = document.getText();
                // Now we write to the file the new content;
                if (vscode.window.activeTextEditor !== undefined) {
                    fs.writeFile(vscode.window.activeTextEditor.document.uri.fsPath, compileFile(text), (err) => {
                        if (err) {
                            let error = vscode.window.createOutputChannel("Latex Compiler Compilation Error");
                            error.appendLine("An error ocurred, no more information available.");
                            error.show();
                            (0, process_1.exit)();
                        }
                        else {
                            vscode.window.showInformationMessage("File Compiled!");
                        }
                    });
                }
            });
        }
    });
}
exports.activate = activate;
function deactivate() { }
exports.deactivate = deactivate;
//# sourceMappingURL=extension.js.map